File: src/models/message.ts
import {
  Entity, BaseEntity,
  Column, PrimaryGeneratedColumn, CreateDateColumn, UpdateDateColumn,
  ManyToOne,
  JoinColumn,
} from "typeorm";
import User from "./user";
import Subscription from "./subscription";
import { Exclude, Expose, Transform } from "class-transformer";

@Entity()
@Exclude()
export default class Message extends BaseEntity {
  constructor(
    owner: User,
    subscription: Subscription,
    title: string,
    abstract: string,
    content: string,
    href?: string | null,
  ) {
    super();
    this.owner = owner;
    this.subscription = subscription;
    this.title = title;
    this.summary = abstract;
    this.content = content;
    this.href = href;
  }

  @PrimaryGeneratedColumn("uuid")
  @Expose()
  public id!: string;

  @Column()
  @Expose()
  public readed: boolean = false;

  @ManyToOne(() => User, (user) => user.messages, {
    eager: true,
  })
  @JoinColumn({ name: "owner_id" })
  @Expose()
  @Transform((value: User) => value.id)
  public owner: User;

  @ManyToOne(() => Subscription, (subscription) => subscription.messages, {
    eager: true,
  })
  @JoinColumn({ name: "subscription_id" })
  @Expose()
  @Transform((value: Subscription) => value.id)
  public subscription: Subscription;

  @Column("text")
  @Expose()
  public title: string;

  @Column("text")
  @Expose()
  public summary: string;

  @Column("text")
  @Expose({ since: 1.1 })
  public content: string;

  @CreateDateColumn({ name: "created_at" })
  @Expose({ name: "created_at" })
  public created_at!: Date;

  @UpdateDateColumn({ name: "updated_at" })
  @Expose({ name: "updated_at" })
  public updated_at!: Date;

  @Expose()
  @Column("varchar", { nullable: true })
  public href?: string | null;
}

File: src/model/message.rs
use super::Subscription;
use chrono::{DateTime, Utc};
use schema::message;
use uuid::Uuid;

#[derive(Queryable, Serialize, Identifiable, Associations)]
#[belongs_to(Subscription)]
#[table_name = "message"]
pub struct Message {
    pub id: Uuid,
    pub readed: bool,
    pub title: String,
    pub summary: String,
    pub content: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    #[serde(rename = "subscription")]
    pub subscription_id: Uuid,
    pub href: Option<String>,
}

#[derive(AsChangeset, Default, Deserialize)]
#[table_name = "message"]
pub struct MessageChangeset {
    pub readed: Option<bool>,
}

#[derive(Serialize)]
pub struct MessageView {
    #[serde(flatten)]
    pub message: Message,
    #[serde(rename = "owner")]
    pub owner_id: Uuid,
}

impl From<(Message, Uuid)> for MessageView {
    fn from((message, owner_id): (Message, Uuid)) -> Self {
        Self { message, owner_id }
    }
}

impl From<(Message, Subscription)> for MessageView {
    fn from((message, subscription): (Message, Subscription)) -> Self {
        Self {
            message,
            owner_id: subscription.owner_id,
        }
    }
}
